### PR TITLE
hdf5: switch to cmake build

### PR DESCRIPTION
**Pull request by sheepforce**
## Description of changes

Switches from autotools to cmake build for hdf5. The autotools build is, among other things, missing the pkg-config files. Their absence in turn breaks the haskell hdf5 packages.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by sheepforce** [comment]
nixpkgs-review lists some packages broken due to changed include paths. I will try adapt them.

File: pkgs/tools/misc/hdf5/default.nix [hunk 1]
```diff
@@ -67,24 +70,15 @@ stdenv.mkDerivation rec {
...
-
-  patches = [
-    # Avoid non-determinism in autoconf build system:
-    # - build time
-    # - build user
-    # - uname -a (kernel version)
-    # Can be dropped once/if we switch to cmake.
-    ./hdf5-more-determinism.patch
-  ];
+  cmakeFlags = lib.optional cppSupport "-DHDF5_BUILD_CPP_LIB=ON"
```
Comment (AndersonTorres) [inline pkgs/tools/misc/hdf5/default.nix:73]: It remembers me to write `cmakeOption`.

**Comment by r-burns** [comment]
> the cmake files are now correctly found by other derivations but the cmake files list the `include` directories of the `hdf5.dev` output, while the cmake files are part of the `hdf.out` output

Yeah, usually cmake files should be in the .dev output. This might be as easy as `moveToOutput [...] "${!outputDev}"` but might get complicated if the cmake files self-reference using absolute paths. Alternatively there might be a way to specify the directory to which cmake files should be installed, so no post-install fixup is needed.

**Comment by r-burns** [comment]
Okay yeah looks like they have a cmake var to specify this:

```diff
--- a/pkgs/tools/misc/hdf5/default.nix
+++ b/pkgs/tools/misc/hdf5/default.nix
@@ -78,6 +78,7 @@ stdenv.mkDerivation rec {
     ++ lib.optional javaSupport "-DHDF5_BUILD_JAVA=ON"
     ++ lib.optional usev110Api "-DDEFAULT_API_VERSION=v110"
     ++ lib.optionals threadsafe [ "-DDHDF5_ENABLE_THREADSAFE:BOOL=ON" "-DHDF5_BUILD_HL_LIB=OFF" ]
+    ++ [ "-DHDF5_INSTALL_CMAKE_DIR=${placeholder "dev"}/lib/cmake" ]
   ;
 
   postInstall = ''
```

Not sure if lib/cmake is correct here, what were the downstream failures you saw? I'll see if I can get those working.

File: pkgs/tools/misc/hdf5/default.nix (hunk 1 above)
Comment (markuskowa) [reply pkgs/tools/misc/hdf5/default.nix:73]: @AndersonTorres What is `cmakeOption`?

Comment (AndersonTorres) [reply pkgs/tools/misc/hdf5/default.nix:73]: Something like `mesonOption` as defined in `lib/strings.nix` file.

**Comment by sheepforce** [comment]
> what were the downstream failures you saw? I'll see if I can get those working.

Most packages fail with this:

```
CMake Error in opencv_contrib/hdf/CMakeLists.txt:
  Imported target "hdf5-shared" includes non-existent path

    "/nix/store/m17gmgz93igbyi1igyn2475n36c7vzj6-hdf5-1.14.2/include"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.
```

**Comment by sheepforce** [comment]
`nixpkgs-review` look much better with @r-burns suggested change. There is still a bunch of packages failing but none of them in configuration phase. Those failing, failed mostly with API and linking errors unrelated to HDF5, Python modules that depend on multiple versions of the same package, packages that don't have sources available or Cuda failures. Here is the full list of all failing packages: 
[failedPackages.txt](https://github.com/NixOS/nixpkgs/files/12586005/failedPackages.txt)

**Comment by markuskowa** [comment]
> Also I guess this better targets staging @markuskowa ?

Yes staging is better, this is about 600-700 rebuilds and the change is not vital for security or bug fixing.

**Review by markuskowa — APPROVED** [review]
LGTM. nixpkgs-review did not turn up any problems as far as I can see.

**Comment by vcunat** [comment]
I suspect this is what now broke `netcdf` tests on *-darwin:
https://hydra.nixos.org/build/236807182/nixlog/4/tail

**Comment by wegank** [comment]
> I suspect this is what now broke `netcdf` tests on *-darwin: https://hydra.nixos.org/build/236807182/nixlog/4/tail

#262937 should fix this.

**Pull request merged**